Dev

vladdord opened the pull request:
Request only for review it won't be merged

RomanEgel committed:
DeutscheBankProject v1.0

In first version were implemented
1)receiving history of brent oil prices in set range through intrinio
service api
2)simple user interface
3)displaying chart price
4)saving prices in the mySql DB

RomanEgel committed:
hide properties

RomanEgel committed:
hide target

RomanEgel committed:
DeutscheBankProject v1.1

new features:
-parsing html page from xmarkets.db and getting from there current brent
oil price
-displaying real-time price
-web-socket technoligy etc

RomanEgel committed:
price getter interface

RomanEgel committed:
add .gitingore

vladdord committed:
Renamed package to com.twoez
Changed way to obtain latest price from xmarkets.db.com

vladdord committed:
Added firs implementation for PriceDispatcher

vladdord committed:
Restructured project as Maven multi-module project

Added histryuploader module with it's README.md
moved HTMLReaderHelper class into common module

Not sure if I've done whole thing right, so it can be fully changed

vladdord committed:
Changed HistoryUploader's readme due to mighration

vladdord committed:
Fixed pom to work with parent package command

vladdord committed:
Removed historical method from PriceGetter
Now whole project is broken and needs to be reconstructed
Changed PriceDispatcher setCurrentPriceGetter method argument from XMarketsPriceGetter -> PriceGetter
XMarketsPriceGetter setConsumer method now inherited from PriceGetter

RomanEgel committed:
DeutscheBankProject v2.0

In this reliaze some acrhitecture problems were resolved. Also we add
composite primary key for brent oil entity and ability to save real-time
prices in db mysql

RomanEgel committed:
Merge branch 'dev' of https://github.com/RomanEgel/DeutscheBankProject into dev

vladdord committed:
Edited README.md due to changes

Moved SQLConfiguration.java to allow predictor use it as preferences object to connect to database
Changed historyuploader due to serve predictor's needs. Now it creates full historical dataset for predictor to train and fill database with predicted prices
Now history uploader modifies database behaviour making database create every 5 seconds new daily dataset

Implemented predictor using SparkML and LinearRegression model for prediction historical and real-time prices
Predictor takes prices from Apache Kafka message topic which fills by site and returns predictions using Kafka but another topic
Site now has functionality to send last obtained price via Kafka and takes predictions from it

RomanEgel committed:
new ui

RomanEgel committed:
fixed bug with prediction. now it works fine

vladdord committed:
Fixed predictor behaviour without network connection
Due to it README.md has changes for server.properties file variables

vladdord committed:
Minor fix of kafkf-clients dependency in module site

bedrin commented on README.md:
```diff
@@ -1 +1,30 @@
+  3. In `server.properties` variables `zookeeper.connect=localhost:2181` and `host.name=localhost` must be set. Last to allow connection to Kafka when network connection is not available.
+  4. After that you can run Zookeeper as `./zookeeper-server-start.sh -daemon ../config/zookeeper.properties` from terminal.
+  5. Now you need to create 2 topics in Kafka. For that use: `./kafka-topics.sh --create --zookeeper localhost:2181 --replication-factor 1 --partitions 1 --topic [topic_name] --config retention.ms=2000` where topic names are **current_price** and **predicted_prices**. Configuration variable `retention.ms` sets frequency of topic purging.
+  6. Now you are ready to run Kafka as message broker for whole system. Use: `./kafka-server-start.sh -daemon ../config/server.properties`
```
I think `-daemon` not required here. Also it should be done before you attempt to create topics - swap with step above

vladdord replied:
True, we figured it yesterday)

bedrin commented on README.md:
```diff
@@ -1 +1,30 @@
+  4. **predictor**: prediction engine for historical and real-time processing
+
+## Requirements
+
```
I would add `mysql` here as well.

bedrin commented on historyuploader/src/main/java/com/twoez/historyuploader/MySqlUploader.java:
```diff
@@ -0,0 +1,173 @@
+
+public class MySqlUploader implements DBUploader{
+
+    private Logger logger = LoggerFactory.getLogger(MySqlUploader.class);
```
static final

bedrin commented on historyuploader/src/main/java/com/twoez/historyuploader/Uploader.java:
```diff
@@ -0,0 +1,59 @@
+@CommandLine.Command(name = "Uploader", mixinStandardHelpOptions = true, version = "0.1")
+public class Uploader implements Callable<Integer> {
+
+    private Logger logger = LoggerFactory.getLogger(Uploader.class);
```
static final

bedrin commented on historyuploader/src/main/java/com/twoez/historyuploader/Uploader.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Callable;
+@CommandLine.Command(name = "Uploader", mixinStandardHelpOptions = true, version = "0.1")
```
Like it

bedrin commented on historyuploader/src/main/java/com/twoez/historyuploader/Uploader.java:
```diff
@@ -0,0 +1,59 @@
+    String dbType;
+
+    @CommandLine.Option(names = {"-n", "--db_name"},required = true, description = "Name of database where to store prices")
+
```
Redundant new lines - please remove. Looks awful when you open the class for the first time :)

bedrin commented on README.md:
```diff
@@ -1 +1,30 @@
+
+## Run
+
+  Now module running order is sufficient so first run **historyuploader** then **predictor** and only then **site**. But breaking order will not cause any problems if only **historyuploader** initialized database.
```
Didn't get from description if order is important or not :)

In general README.md should allow a new developer to run the project in a few minutes (or even seconds) so it should be super straightforward

bedrin commented on historyuploader/src/main/java/com/twoez/historyuploader/MySqlUploader.java:
```diff
@@ -0,0 +1,173 @@
+
+        logger.info("Procedure dropped");
+        logger.info("Creating procedure");
+        String procedureCreateStatement = "create procedure create_dataset_table() READS SQL DATA begin drop table if exists new_table; drop table if exists next_price_table; create table new_table(id int AUTO_INCREMENT,price double,next_price double, PRIMARY KEY(id)); insert into new_table (price) select price from brent_oil where source = 'RealTimeXMarkets' or source = 'HistoryProfinance' order ...
```
![](https://cnet1.cbsistatic.com/img/M_JRv2wKCCjMKVvYk0xIjOfSk-E=/830x467/2018/09/07/f2067f46-fb2f-445c-9269-eb28aa8c9da2/elon.jpg)

bedrin commented on historyuploader/src/main/java/com/twoez/historyuploader/Uploader.java:
```diff
@@ -0,0 +1,59 @@
+
+    private Logger logger = LoggerFactory.getLogger(Uploader.class);
+
+    @CommandLine.Option(names = {"-d","--db_type"}, required = true, description = "Type of database where you want to store your prices.\n" + "Currently available: [mysql]")
```
This is unfair - you don't even use it. Make it optional for now

bedrin commented on site/src/main/resources/templates/getCurrent.html:
```diff
@@ -0,0 +1,109 @@
+                return false;
+            }
+
+            window.onbeforeunload = endCon;
```
This bastard ruins the user experience with annoying popups

bedrin commented on common/src/main/java/com/twoez/common/HTMLReaderHelper.java:
```diff
@@ -0,0 +1,137 @@
+        return new StringIterator();
+    }
+
+    public class CharIterator implements java.util.Iterator<Character>{
```
You really need to hate libraries in order to come up with this :) 
What was wrong with BufferedReader ?

vladdord replied:
We thought that buffering policy in BufferedReader might slow down getting price. So we rewrote it. But that was reasonable until we found out that `mark()` and `reset()` methods are not supported by our JDK

bedrin replied:
BufferedReader should be fine in your case.
One scenario where it would be ineffective is when the whole HTML is formatted into single line, but it is not the case with xmarkets. Premature optimization is evil

bedrin commented on site/src/main/resources/templates/uploadValues.html:
```diff
@@ -0,0 +1,183 @@
+
+                var chart = new google.visualization.AreaChart(document.getElementById('curve_chart'));
+
+                chart.draw(data, options);
```
As discussed - would be cool to have both predicted and historical prices on a single chart.
I really think it would be a killer feature

bedrin reviewed (approved):
Great work! I left a few minor comments.
It deserves to be merged into master 👍

vladdord committed:
Edited code due to comments in last pull request
Modified README.md to be more clear about how to start project
Modified historyuploader.Uploader and MySqlUplader classes due to comments
Small changes in HistoryCoverage class
Tried to fix NullPointerException problem in Iterator's of HTMLReaderHelper

RomanEgel committed:
added dynamic charts

vladdord committed:
Another try to fix NullPointerException in HTMLReaderHelper
Removed debug output from listeners in site

RomanEgel committed:
added history prediction chart and fixed some bugs

RomanEgel committed:
Merge branch 'dev' of https://github.com/RomanEgel/DeutscheBankProject into dev
